### PR TITLE
Don't use longjmp and check for the error indicator on every rule

### PR DESCRIPTION
**Pull request by pablogsal**

### CODE REVIEW

File: Parser/pegen/pegen.c [hunk 1]
```diff
@@ -60,7 +60,8 @@ _PyPegen_new_identifier(Parser *p, char *n)
                          "%.200s",
                          _PyType_Name(Py_TYPE(id2)));
             Py_DECREF(id2);
-            longjmp(p->error_env, 1);
+            p->error_indicator = 1;
+            return NULL;
```
Comment (gvanrossum) [inline Parser/pegen/pegen.c:64]: Why is it set here but not at earlier `return NULL` lines in this function?

File: Parser/pegen/pegen.h [hunk 2]
```diff
@@ -117,7 +116,7 @@ CHECK_CALL(Parser *p, void *result)
 {
     if (result == NULL) {
         assert(PyErr_Occurred());
-        longjmp(p->error_env, 1);
+        p->error_indicator=1;
```
Comment (gvanrossum) [inline Parser/pegen/pegen.h:119]: PEP 7

File: Tools/peg_generator/pegen/c_generator.py [hunk 3]
```diff
@@ -321,6 +322,10 @@ def _handle_default_rule_body(self, node: Rule, rhs: Rhs, result_type: str) -> N
         memoize = self._should_memoize(node)
 
         with self.indent():
+            self.print("if (p->error_indicator){")
```
Comment (gvanrossum) [inline Tools/peg_generator/pegen/c_generator.py:325]: PEP 7

File: Tools/peg_generator/pegen/c_generator.py [hunk 4]
```diff
@@ -349,6 +354,10 @@ def _handle_loop_rule_body(self, node: Rule, rhs: Rhs) -> None:
         is_repeat1 = node.name.startswith("_loop1")
 
         with self.indent():
+            self.print("if (p->error_indicator){")
```
Comment (gvanrossum) [inline Tools/peg_generator/pegen/c_generator.py:357]: PEP 7

**Comment by gvanrossum** [comment]
Sorry the line I linked to isn't clear in the review. It's at line 10109 in the generated parse.c (and at many other places too). I believe generated at L387 in c_generator.py. (By `self.out_of_memory_return()`.)

**Comment by gvanrossum** [comment]
There's also a leak in generated code like this. The `return NULL;` following `p->error_indicator = 1;` leaks `children`.

```
// _loop0_4: ';' small_stmt
static asdl_seq *
_loop0_4_rule(Parser *p)
{
    if (p->error_indicator){
        return NULL;
    }
    void *res = NULL;
    int mark = p->mark;
    int start_mark = p->mark;
    void **children = PyMem_Malloc(sizeof(void *));
    if (!children) {
        PyErr_Format(PyExc_MemoryError, "Parser out of memory");
        return NULL;
    }
    ssize_t children_capacity = 1;
    ssize_t n = 0;
    { // ';' small_stmt
        stmt_ty elem;
        void *literal;
        while (
            (literal = _PyPegen_expect_token(p, 13))
            &&
            (elem = small_stmt_rule(p))
        )
        {
            res = elem;
            if (res == NULL && PyErr_Occurred()) {
                p->error_indicator = 1;
                return NULL;
            }
            if (n == children_capacity) {
                children_capacity *= 2;
                children = PyMem_Realloc(children, children_capacity*sizeof(void *));
                if (!children) {
                    PyErr_Format(PyExc_MemoryError, "realloc None");
                    return NULL;
                }
            }
            children[n++] = res;
            mark = p->mark;
        }
        p->mark = mark;
    }
    asdl_seq *seq = _Py_asdl_seq_new(n, p->arena);
    if (!seq) {
        PyErr_Format(PyExc_MemoryError, "asdl_seq_new _loop0_4");
        return NULL;
    }
    for (int i = 0; i < n; i++) asdl_seq_SET(seq, i, children[i]);
    PyMem_Free(children);
    _PyPegen_insert_memo(p, start_mark, _loop0_4_type, seq);
    return seq;
}
```

**Comment by pablogsal** [comment]
After 023ed5e now the function you mention looks like this:

```C
_loop0_4_rule(Parser *p)
{
    if (p->error_indicator) {
        return NULL;
    }
    void *res = NULL;
    int mark = p->mark;
    int start_mark = p->mark;
    void **children = PyMem_Malloc(sizeof(void *));
    if (!children) {
        PyErr_Format(PyExc_MemoryError, "Parser out of memory");
        return NULL;
    }
    ssize_t children_capacity = 1;
    ssize_t n = 0;
    { // ';' small_stmt
        stmt_ty elem;
        void *literal;
        while (
            (literal = _PyPegen_expect_token(p, 13))
            &&
            (elem = small_stmt_rule(p))
        )
        {
            res = elem;
            if (res == NULL && PyErr_Occurred()) {
                p->error_indicator = 1;
                PyMem_Free(children);
                return NULL;
            }
            if (n == children_capacity) {
                children_capacity *= 2;
                children = PyMem_Realloc(children, children_capacity*sizeof(void *));
                if (!children) {
                    PyErr_Format(PyExc_MemoryError, "realloc None");
                    return NULL;
                }
            }
            children[n++] = res;
            mark = p->mark;
        }
        p->mark = mark;
    }
    asdl_seq *seq = _Py_asdl_seq_new(n, p->arena);
    if (!seq) {
        PyErr_Format(PyExc_MemoryError, "asdl_seq_new _loop0_4");
        PyMem_Free(children)
        return NULL;
    }
    for (int i = 0; i < n; i++) asdl_seq_SET(seq, i, children[i]);
    PyMem_Free(children);
    _PyPegen_insert_memo(p, start_mark, _loop0_4_type, seq);
    return seq;
}
```

File: Parser/pegen/pegen.c [hunk 5]
```diff
@@ -72,6 +72,10 @@ _PyPegen_new_identifier(Parser *p, char *n)
         return NULL;
```
Comment (gvanrossum) [inline Parser/pegen/pegen.c:72]: Why no goto error here?

File: Tools/peg_generator/pegen/c_generator.py [hunk 6]
```diff
@@ -174,11 +174,14 @@ def call_with_errorcheck_goto(self, call_text: str, goto_target: str) -> None:
         self.print(f"}}")
 
     def out_of_memory_return(
-        self, expr: str, returnval: str, message: str = "Parser out of memory"
+        self, expr: str, returnval: str, message: str = "Parser out of memory",
+        cleanup_code=None
```
Comment (gvanrossum) [inline Tools/peg_generator/pegen/c_generator.py:178]: Might want to re-run black?

**Pull request merged**